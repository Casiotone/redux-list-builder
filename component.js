import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import {
  LIST_BUILDER_ADD_ITEM,
  LIST_BUILDER_REMOVE_ITEM,
} from './actions';

import './index.scss';

class ListBuilder extends Component {

  constructor(props) {
    super(props);
    this.state = { value: '' };
  }

  validateInput(value) {
    const { validation } = this.props;
    return validation ? validation(value) : true;
  }

  handleSubmit() {
    const { value } = this.state;
    if(this.validateInput(value.trim())) {
      this.props.dispatch({
        type: LIST_BUILDER_ADD_ITEM,
        value,
      });
      this.setState({ value: '' });
    }
  }

  render() {

    const {
      dispatch,
      list,
      inputLabel,
      buttonLabel,
      formClassName,
      listClassName,
    } = this.props;

    return (

      <form onSubmit={(e) => {
        e.preventDefault();
        this.handleSubmit();
      }} className={formClassName || 'redux-list-builder'}>
        <div className={'form-group form-group-combined'}>
          <label htmlFor={'listItemValue'}>{inputLabel || 'Item value'}</label>
          <div className={'form-control-wrapper'}>
            <input
              className={'form-control'}
              name={'listItemValue'}
              value={this.state.value}
              onChange={(e) => {this.setState({ value: e.target.value });}}
              onBlur={() => {this.handleSubmit();}}
            />
            <button
              type={'button'}
              className={'btn btn-primary'}
            >{buttonLabel || 'Add'}</button>
          </div>

          <ul className={listClassName || ''}>
            {list.map((item, index) => {
              return (
                <li>
                    <span className={'list-item-label'}>{item}</span>
                    <button
                      className={'btn btn-transparent list-item-remove'}
                      onClick={() => {
                        dispatch({
                          type: LIST_BUILDER_REMOVE_ITEM,
                          value: index,
                        });
                      }}
                    >x</button>
                </li>
              );
            })}
          </ul>

        </div>
      </form>

    );

  }

}

ListBuilder.propTypes = {
  list: PropTypes.array.isRequired,
  dispatch: PropTypes.func,
  validation: PropTypes.func,
  buttonLabel: PropTypes.string,
  inputLabel: PropTypes.string,
  formClassName: PropTypes.string,
  listClassName: PropTypes.string,
};

export default connect(
  (state) => ({
    list: state.listBuilder,
  })
)(ListBuilder);
