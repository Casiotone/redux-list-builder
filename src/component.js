import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
    LIST_BUILDER_CREATE_LIST,
    LIST_BUILDER_ADD_ITEM,
    LIST_BUILDER_REMOVE_ITEM,
} from './redux';

const defaultStyles = {
    wrapper: {
        display: 'flex',
        flexDirection: 'row',
        flexWrap: 'wrap',
    },
    listItem: {
        backgroundColor: '#333',
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        margin: '2px',
    },
    itemLabel: {
        padding: '0 5px',
        color: '#fff',
    },
    itemRemove: {
        color: '#333',
        margin: '-3px 0 0',
    },
};

export class ListBuilder extends Component {
    constructor(props) {
        super(props);
        const { dispatch, identifier } = props;
        dispatch({
            type: LIST_BUILDER_CREATE_LIST,
            payload: { identifier },
        });
        this.state = { value: '' };
    }

    handleSubmit() {
        const { dispatch, validation, identifier } = this.props;
        const { value } = this.state;
        if (validation(value.trim())) {
            dispatch({
                type: LIST_BUILDER_ADD_ITEM,
                payload: {
                    identifier,
                    item: {
                        id: `${new Date().valueOf()}${Math.random().toFixed(16).substring(2)}`,
                        value,
                    },
                },
            });
            this.setState({ value: '' });
        }
    }

    render() {
        const {
            identifier,
            dispatch,
            list,
            inputLabel,
            buttonLabel,
            itemRemoveSymbol,
            classNames: {
                formClassName,
                labelClassName,
                wrapperClassName,
                inputClassName,
                submitButtonClassName,
                listClassName,
                itemClassName,
                itemLabelClassName,
                itemRemoveClassName,
            },
        } = this.props;

        const {
            wrapper,
            listItem,
            itemLabel,
            itemRemove,
        } = defaultStyles;

        return (
            <form
                onSubmit={(e) => {
                    e.preventDefault();
                    this.handleSubmit();
                }}
                className={formClassName}
            >
                <div className={wrapperClassName} style={!wrapperClassName && { ...wrapper }}>
                    <label htmlFor={'listItemValue'} className={labelClassName}>
                        {inputLabel}
                        <input
                            className={inputClassName}
                            id={'listItemValue'}
                            value={this.state.value}
                            onChange={(e) => { this.setState({ value: e.target.value }); }}
                        />
                        <button
                            type={'submit'}
                            className={submitButtonClassName}
                        >{buttonLabel}
                        </button>
                    </label>
                </div>
                <ul className={listClassName}>
                    {list.map(item => (
                        <li key={item.id} className={itemClassName} style={!itemClassName && { ...listItem }}>
                            <span className={itemLabelClassName} style={!itemLabelClassName && { ...itemLabel }}>{item.value}</span>
                            <button
                                className={itemRemoveClassName}
                                style={!itemRemoveClassName && { ...itemRemove }}
                                onClick={() => {
                                    dispatch({
                                        type: LIST_BUILDER_REMOVE_ITEM,
                                        payload: {
                                            id: item.id,
                                            identifier,
                                        },
                                    });
                                }}
                            >{itemRemoveSymbol}
                            </button>
                        </li>
                    ))}
                </ul>
            </form>
        );
    }
}

ListBuilder.defaultProps = {
    validation: value => !!value.match(/[^\s]/),
    buttonLabel: 'Add',
    inputLabel: 'Item value',
    itemRemoveSymbol: 'X',
    classNames: {
        formClassName: '',
        labelClassName: '',
        wrapperClassName: '',
        inputClassName: '',
        submitButtonClassName: '',
        listClassName: '',
        itemClassName: '',
        itemLabelClassName: '',
        itemRemoveClassName: '',
    },
};

ListBuilder.propTypes = {
    identifier: PropTypes.string.isRequired,
    list: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string,
        value: PropTypes.string,
    })).isRequired,
    dispatch: PropTypes.func.isRequired,
    validation: PropTypes.func,
    buttonLabel: PropTypes.string,
    inputLabel: PropTypes.string,
    itemRemoveSymbol: PropTypes.string,
    classNames: PropTypes.shape({
        formClassName: PropTypes.string,
        labelClassName: PropTypes.string,
        wrapperClassName: PropTypes.string,
        inputClassName: PropTypes.string,
        submitButtonClassName: PropTypes.string,
        listClassName: PropTypes.string,
        itemClassName: PropTypes.string,
        itemLabelClassName: PropTypes.string,
        itemRemoveClassName: PropTypes.string,
    }),
};

export default connect(
    (state, props) => ({
        list: state.listBuilder[props.identifier] || [],
    }),
)(ListBuilder);
