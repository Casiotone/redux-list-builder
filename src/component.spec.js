import React from 'react';
import { shallow } from 'enzyme';
import { getProps } from '../test/helpers';
import { ListBuilder } from './component';

const props = {
    identifier: 'listID',
    dispatch: jest.fn(),
    list: [],
    validation: value => !!value.match(/[^\s]/),
    buttonLabel: 'Add',
    inputLabel: 'Item value',
    itemRemoveSymbol: 'X',
    classNames: {
        formClassName: '',
        labelClassName: '',
        wrapperClassName: '',
        inputClassName: '',
        submitButtonClassName: '',
        listClassName: '',
        itemClassName: '',
        itemLabelClassName: '',
        itemRemoveClassName: '',
    },
};

describe('ListBuilder component', () => {
    const wrapper = shallow(<ListBuilder {...getProps(props)} />);

    it('should render', () => {
        expect(wrapper.find('form')).toMatchSnapshot();
    });
});
