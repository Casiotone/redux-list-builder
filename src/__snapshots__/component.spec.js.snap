// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBuilder component should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): ShallowWrapper {
    "length": 1,
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <ListBuilder
      buttonLabel="Add"
      classNames={
        Object {
          "formClassName": "",
          "inputClassName": "",
          "itemClassName": "",
          "itemLabelClassName": "",
          "itemRemoveClassName": "",
          "labelClassName": "",
          "listClassName": "",
          "submitButtonClassName": "",
          "wrapperClassName": "",
        }
      }
      dispatch={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "payload": Object {
                  "identifier": "listID",
                },
                "type": "redux-flash-message/LIST_BUILDER_CREATE_LIST",
              },
            ],
          ],
        }
      }
      identifier="listID"
      inputLabel="Item value"
      itemRemoveSymbol="X"
      list={Array []}
      validation={[Function]}
    />,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className=""
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "flexWrap": "wrap",
              }
            }
          >
            <label
              className=""
              htmlFor="listItemValue"
            >
              Item value
              <input
                className=""
                id="listItemValue"
                onChange={[Function]}
                value=""
              />
              <button
                className=""
                type="submit"
              >
                Add
              </button>
            </label>
          </div>,
          <ul
            className=""
          />,
        ],
        "className": "",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <label
              className=""
              htmlFor="listItemValue"
            >
              Item value
              <input
                className=""
                id="listItemValue"
                onChange={[Function]}
                value=""
              />
              <button
                className=""
                type="submit"
              >
                Add
              </button>
            </label>,
            "className": "",
            "style": Object {
              "display": "flex",
              "flexDirection": "row",
              "flexWrap": "wrap",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Item value",
                <input
                  className=""
                  id="listItemValue"
                  onChange={[Function]}
                  value=""
                />,
                <button
                  className=""
                  type="submit"
                >
                  Add
                </button>,
              ],
              "className": "",
              "htmlFor": "listItemValue",
            },
            "ref": null,
            "rendered": Array [
              "Item value",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "",
                  "id": "listItemValue",
                  "onChange": [Function],
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Add",
                  "className": "",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Add",
                "type": "button",
              },
            ],
            "type": "label",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": "form",
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className=""
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                }
              }
            >
              <label
                className=""
                htmlFor="listItemValue"
              >
                Item value
                <input
                  className=""
                  id="listItemValue"
                  onChange={[Function]}
                  value=""
                />
                <button
                  className=""
                  type="submit"
                >
                  Add
                </button>
              </label>
            </div>,
            <ul
              className=""
            />,
          ],
          "className": "",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <label
                className=""
                htmlFor="listItemValue"
              >
                Item value
                <input
                  className=""
                  id="listItemValue"
                  onChange={[Function]}
                  value=""
                />
                <button
                  className=""
                  type="submit"
                >
                  Add
                </button>
              </label>,
              "className": "",
              "style": Object {
                "display": "flex",
                "flexDirection": "row",
                "flexWrap": "wrap",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Item value",
                  <input
                    className=""
                    id="listItemValue"
                    onChange={[Function]}
                    value=""
                  />,
                  <button
                    className=""
                    type="submit"
                  >
                    Add
                  </button>,
                ],
                "className": "",
                "htmlFor": "listItemValue",
              },
              "ref": null,
              "rendered": Array [
                "Item value",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "",
                    "id": "listItemValue",
                    "onChange": [Function],
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Add",
                    "className": "",
                    "type": "submit",
                  },
                  "ref": null,
                  "rendered": "Add",
                  "type": "button",
                },
              ],
              "type": "label",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "",
            },
            "ref": null,
            "rendered": Array [],
            "type": "ul",
          },
        ],
        "type": "form",
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className=""
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "flexWrap": "wrap",
            }
          }
        >
          <label
            className=""
            htmlFor="listItemValue"
          >
            Item value
            <input
              className=""
              id="listItemValue"
              onChange={[Function]}
              value=""
            />
            <button
              className=""
              type="submit"
            >
              Add
            </button>
          </label>
        </div>,
        <ul
          className=""
        />,
      ],
      "className": "",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <label
            className=""
            htmlFor="listItemValue"
          >
            Item value
            <input
              className=""
              id="listItemValue"
              onChange={[Function]}
              value=""
            />
            <button
              className=""
              type="submit"
            >
              Add
            </button>
          </label>,
          "className": "",
          "style": Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item value",
              <input
                className=""
                id="listItemValue"
                onChange={[Function]}
                value=""
              />,
              <button
                className=""
                type="submit"
              >
                Add
              </button>,
            ],
            "className": "",
            "htmlFor": "listItemValue",
          },
          "ref": null,
          "rendered": Array [
            "Item value",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "",
                "id": "listItemValue",
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add",
                "className": "",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Add",
              "type": "button",
            },
          ],
          "type": "label",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className=""
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "flexWrap": "wrap",
              }
            }
          >
            <label
              className=""
              htmlFor="listItemValue"
            >
              Item value
              <input
                className=""
                id="listItemValue"
                onChange={[Function]}
                value=""
              />
              <button
                className=""
                type="submit"
              >
                Add
              </button>
            </label>
          </div>,
          <ul
            className=""
          />,
        ],
        "className": "",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <label
              className=""
              htmlFor="listItemValue"
            >
              Item value
              <input
                className=""
                id="listItemValue"
                onChange={[Function]}
                value=""
              />
              <button
                className=""
                type="submit"
              >
                Add
              </button>
            </label>,
            "className": "",
            "style": Object {
              "display": "flex",
              "flexDirection": "row",
              "flexWrap": "wrap",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Item value",
                <input
                  className=""
                  id="listItemValue"
                  onChange={[Function]}
                  value=""
                />,
                <button
                  className=""
                  type="submit"
                >
                  Add
                </button>,
              ],
              "className": "",
              "htmlFor": "listItemValue",
            },
            "ref": null,
            "rendered": Array [
              "Item value",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "",
                  "id": "listItemValue",
                  "onChange": [Function],
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Add",
                  "className": "",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Add",
                "type": "button",
              },
            ],
            "type": "label",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
